### PR TITLE
Logs limit & log_index bug

### PR DESCRIPTION
**Pull request by tomusdrw**
`eth_getLogs`  API with optional parameter to limit number of results - useful in GUI.
Moved logs to `blockchain` to be able to test it.

Perhaps it would be better to move it to `ethcore_*` API instead of introducing additional parameter?

### CODE REVIEW

File: ethcore/src/blockchain/blockchain.rs [hunk 1]
```diff
@@ -315,6 +320,43 @@ impl BlockProvider for BlockChain {
...
+								entry: log,
+								block_hash: hash.clone(),
+								block_number: number,
+								transaction_hash: hashes.get(index).cloned().unwrap_or_else(H256::default),
+								transaction_index: index,
+								log_index: log_index + i
+							})
+							.collect::<Vec<LocalizedLogEntry>>()
+					})
+					.collect::<Vec<LocalizedLogEntry>>()
```
Comment (rphmeier) [inline ethcore/src/blockchain/blockchain.rs]: I think both these `collect`s are unnecessary

**Comment by tomusdrw** [comment]
Addressed grumbles and fixed a bug with logs indexing. @debris please review tests and logs logic carefully :)

**Comment by rphmeier** [comment]
could you describe the bug and the fix? not sure if I can see it at first glance

File: ethcore/src/blockchain/blockchain.rs [hunk 2]
```diff
@@ -315,6 +320,52 @@ impl BlockProvider for BlockChain {
 			.map(|b| b as BlockNumber)
 			.collect()
 	}
+
+	fn logs<F>(&self, mut blocks: Vec<BlockNumber>, matches: F, limit: Option<usize>) -> Vec<LocalizedLogEntry>
+		where F: Fn(&LogEntry) -> bool, Self: Sized {
+		// sort in reverse order
+		blocks.sort();
+		blocks.reverse();
```
Comment (rphmeier) [inline ethcore/src/blockchain/blockchain.rs]: `sort_by` and `Ordering::reverse` will make this a bit more efficient.

**Comment by tomusdrw** [comment]
log_index bug:
https://github.com/ethcore/parity/blob/master/ethcore/src/client/client.rs#L983

Here we first increment `log_index` and then add `i` when creating `LocalizedLogEntries` - it means that we will never get `LocalizedLogEntry` with `log_index==0`.

**Pull request merged**